src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java

```diff
@@ -444,13 +457,369 @@ protected boolean checkCgIcrRules(String subAccountTypeCode, SubAccountGlobalDet
+        String subAccountKeyFormat = "%s-%s-%s";
+        Set<String> subAccountKeys = new HashSet<>();
+        boolean success = true;
+        
+        int i = 0;
+        for (SubAccountGlobalNewAccountDetail newAccountDetail: subAccountGlobal.getSubAccountGlobalNewAccountDetails()) {
+            String subAccountNumber = subAccountGlobal.isApplyToAllNewSubAccounts()
+                    ? commonSubAccountNumber : newAccountDetail.getSubAccountNumber();
+            String subAccountKey = String.format(subAccountKeyFormat, newAccountDetail.getChartOfAccountsCode(),
+                    newAccountDetail.getAccountNumber(), subAccountNumber);
+            if (!subAccountKeys.add(subAccountKey)) {
+                putFieldErrorForDuplicateOrExistingSubAccount(
+                        CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NEW_SUB_ACCOUNT_DUPLICATE, subAccountKey, i);
+                success = false;
+            } else {
+                success &= checkSubAccountDoesNotExist(newAccountDetail, subAccountNumber, subAccountKey, i);
+            }
+            i++;
+        }
+        
+        return success;
+    }
+    
+    protected boolean checkSubAccountDoesNotExist(
+            SubAccountGlobalNewAccountDetail newAccountDetail, String subAccountNumber, String subAccountKey, int index) {
+        SubAccount subAccount = getSubAccountService().getByPrimaryId(
+                newAccountDetail.getChartOfAccountsCode(), newAccountDetail.getAccountNumber(), subAccountNumber);
+        if (ObjectUtils.isNotNull(subAccount)) {
+            putFieldErrorForDuplicateOrExistingSubAccount(
+                    CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NEW_SUB_ACCOUNT_EXISTS, subAccountKey, index);
+            return false;
+        }
+        return true;
+    }
+    
+    protected void putFieldErrorForDuplicateOrExistingSubAccount(String errorConstant, String subAccountKey, int index) {
+        String fullPropertyName = buildListObjectPropertyPath(
+                CUKFSPropertyConstants.SUB_ACCOUNT_GLOBAL_NEW_ACCOUNT_DETAILS, KFSPropertyConstants.SUB_ACCOUNT_NUMBER, index);
+        putFieldError(fullPropertyName, errorConstant, subAccountKey);
+    }
+    
+    protected boolean checkContractsAndGrantsSetupForNonEmptyNewAccountsList(SubAccountGlobal subAccountGlobal) {
+        boolean success = true;
+        boolean allAccountsAreForContractsAndGrants = true;
+        boolean atLeastOneAccountIsForContractsAndGrants = false;
+        
+        for (SubAccountGlobalNewAccountDetail newAccountDetail : subAccountGlobal.getSubAccountGlobalNewAccountDetails()) {
+            boolean currentAccountIsForContractsAndGrants = accountIsForContractsAndGrants(newAccountDetail);
+            allAccountsAreForContractsAndGrants &= currentAccountIsForContractsAndGrants;
+            atLeastOneAccountIsForContractsAndGrants |= currentAccountIsForContractsAndGrants;
```

Comment:
I'm still concerned this is confusing.
How about creating integers for cgAccountCount and nonCgAccountCount.  In the for loop increment the counters as appropriate.  Then below check to see if both integers are gerater than zero, if so, error.  Then in the else statement do as you've currently coded.

Reply:
Good idea, I'll work on incorporating counters to see if that makes the logic clearer.